python/lsst/daf/butler/datastores/sasquatchDatastore.py

```diff
@@ -0,0 +1,885 @@
+            identifierFields=identifierFields,
+            extraFields=extraFields,
+        )
+
+        headers = {"content-type": "application/vnd.kafka.avro.v2+json"}
+        data: dict[str, Any] = dict()
+        partialUpload = False
+
+        for metric, record in metricRecords.items():
+            # create the kafka topic if it does not already exist
+            if not self._create_topic(metric):
+                log.error("Topic not created, skipping dispatch of %s", metric)
+                continue
+            recordValue = record[0]["value"]
+            # Generate schemas for each record
+            data["value_schema"], schemaTrimmed = self._generateAvroSchema(metric, recordValue)
+            data["records"] = record
+
+            if schemaTrimmed:
+                partialUpload = True
+
+            r = requests.post(f"{self.url}/topics/{self.namespace}.{metric}", json=data, headers=headers)
+
+            if r.status_code == 200:
+                log.debug("Succesfully sent data for metric %s", metric)
+            else:
+                log.error(
+                    "There was a problem submitting the metric %s: %s, %s", metric, r.status_code, r.json()
+                )
+                partialUpload = True
+
+        if partialUpload or recordsTrimmed:
+            raise SasquatchDispatchPartialFailure("One or more records may not have been uploaded entirely")
+
+    def dispatchRef(
+        self,
+        bundle: MetricMeasurementBundle,
+        ref: DatasetRef,
+        timestamp: datetime.datetime | None = None,
+        extraFields: Mapping | None = None,
+        datasetIdentifier: str | None = None,
+    ) -> None:
+        """Dispatch a `MetricMeasurementBundle` to Sasquatch with a known
+        `DatasetRef`.
+
+        Parameters
+        ----------
+        bundle : `~lsst.analysis.tools.MetricMeasurementBundle`
+            The bundle containing metric values to upload
+        ref : `DatasetRef`
+            The `Butler` dataset ref corresponding to the input
+            `MetricMeasurementBundle`.
+        timestamp : `str` or `None`
+            The timestamp to be associated with the measurements in the ingress
+            database. If this value is None, timestamp will be set by the run
+            time or current time.
+        extraFields: `Mapping` or None
+            Extra mapping keys and values that will be added as fields to the
+            dispatched record if not None.
+        datasetIdentifier : `str` or None
+            A string which will be used in creating unique identifier tags. If
+            None, a default value will be inserted.
+
+        Raises
+        ------
+        SasquatchDispatchPartialFailure
+            Raised if there were any errors in dispatching a bundle.
+        """
+        # Parse the relevant info out of the dataset ref.
+        serializedRef = ref.to_simple()
+        id = serializedRef.id
+        if serializedRef.run is None:
+            run = ""
```

Comment:
```suggestion
            run = "<unknown>"
```
? Since this should be impossible (and soon mypy will tell you this).